src/phoenix/experiments/types.py

```diff
@@ -74,33 +97,77 @@ def from_dict(cls, obj: Mapping[str, Any]) -> Example:
             updated_at=obj["updated_at"],
         )
 
+    def __repr__(self) -> str:
+        spaces = " " * 4
+        name = self.__class__.__name__
+        identifiers = [f'{spaces}id="{self.id}",']
+        contents = [
+            spaces
+            + f"{k}="
+            + json.dumps(
+                _shorten(v),
+                ensure_ascii=False,
+                sort_keys=True,
+                indent=len(spaces),
+            ).replace("\n", f"\n{spaces}")
+            + ","
+            for k in ("input", "output", "metadata")
+            if (v := getattr(self, k, None))
+        ]
+        return "\n".join([f"{name}(", *identifiers, *contents, ")"])
+
 
 @dataclass(frozen=True)
 class Dataset:
     id: DatasetId
     version_id: DatasetVersionId
-    examples: Sequence[Example]
+    examples: Mapping[ExampleId, Example] = field(repr=False, default_factory=dict)
```

Comment:
Where does it help to have a dictionary type rather than a list?

Reply:
Looks like the main reason is to make `Dataset` indexable.

Reply:
1. it enforces the uniqueness
2. easier for lookups: i needed to print the example in the `__repr__` for the experiment run objects to help ppl write their eval functions

Reply:
Makes sense. I feel like constructing the dictionary and validating uniqueness should happen in `__post_init__` so it's encapsulated/ not the responsibility of the caller.

Reply:
Same for `RanExperiment`.

Reply:
true, except the type checker won't be happy about it